getting_started_modules.md

```diff
@@ -0,0 +1,283 @@
+When doing so you will get the error "Error: Missing required argument The argument "address" is required, but was not set.". 
+This is because we are using a resource that deploys a nomad job, but nowhere in our terraform files have we defined _ what_ Nomad to use. 
+At the moment no Nomad is known. This is where [providers]() come into the picture. They are providers for the resources we are using, and in our case we need to define a [Nomad provider](). 
+We could do this in either of our `main.tf` files, but if we do it in our module's `main.tf` it will be very difficult for anyone to use our module, because the Nomad's address is predefined. 
+Instead we should include what nomad to use in our example´s `main.tf`, which is simply an example of how to use the module, meaning anyone else wanting to use the module could supply their own nomad when using the module. 
+To supply a provider add the lines below to your `example/main.tf` file:
+
+```hcl-terraform
+provider "nomad" {
+  address = "http://127.0.0.1:4646"
+}
+```
+
+we have now told terraform what Nomad we want to use. 
+Try running your terraform code with `terraform init` (we need to load the nomad provider), `terraform plan` (this time it should succeed), then lastly `terraform apply`, which will execute the plan. 
+Go to `localhost:4646` to check if the nomad-job has started running, if it has, congratulations, you 
+have made your first working terraform module!
+
+
+
+##### b. Using Ansible To Run the Code in the Previous Step On Startup
+In [getting_started_vagrantbox.md](/getting_started_vagrantbox.md) it was mentioned that ALL ansible tasks put inside `dev/ansible/` will be run when the box starts. 
+We can use
+ this to automatically start our module when we run `make up`. 
+ The ansible code for running terraform code is below. 
+ Add this to `run-terraform.yml` or another aptly named file.
+
+```yml
+- name: Terraform
+  terraform:
+    project_path: ../../example
+    force_init: true
+    state: present
+  register: terraform
+
+- name: Terraform stdout
+  debug:
+    msg: "{{terraform.stdout}}"
+```
+
+
+#### 5. Making The Nomad Job More Dynamic With Terraform Variables
+At this point we have a "hardcoded" nomad job. 
+In the example there isn't really many variables that will benefit from being more dynamic, but we'll demonstrate using the image-name in this case. 
+We want to use this to let the user choose their own image in our module. Create a `variables.tf` in the root-directory, then add the lines below
+
+```hcl-terraform
+variable "docker_image" {
+  type        = string
+  description = "docker image to use in module"
+}
+```
+ 
+Now try to run this module like you've done earlier. 
+At this point you'll get an error TODO:ADD ACTUAL ERROR MSG. 
+This is because there is no code anywhere defining `container_name`. 
+
+TODO: FINISH THIS SECTION
```

Comment:
```suggestion
How dynamic the nomad job suppose to be is defined by the needs of the end-user who are going to use this module. There are two main methods that we recommend.

##### Method 1 - extract dynamic data into variables.tf

We advise starting simple with doing extraction of potentially dynamic data in nomad job, such as:
- job name
- list of data centers
- namespace (enterprise only)
- variables of [update stanza](https://www.nomadproject.io/docs/job-specification/update)
- [service name stanza](https://www.nomadproject.io/docs/job-specification/service#name)
- env variables in [template stanza](https://www.nomadproject.io/docs/job-specification/template)
- docker image in [task -> config -> image](https://www.nomadproject.io/docs/drivers/docker)
- ...

> Extracting variables is a process, so you do not need to extract all dynamic data at once. Try to prioritize user needs.

##### Method 2 - conditional rendering of nomad job
You can add a variable in `variables.tf` to define a condition.
This variable will be used to control if-statement how to render the nomad job.

Example:  

`variables.tf`

/```hcl
variable "use_canary" {
  type = bool
  description = "Uses canary deployment for Presto"
  default = false
}
/```

`nomad-job.hcl`
/```hcl
...
update {
  max_parallel      = 1
  health_check      = "checks"
  min_healthy_time  = "10s"
  healthy_deadline  = "12m"
  progress_deadline = "15m"
  %{ if use_canary }
  canary            = 1
  auto_promote      = true
  auto_revert       = true
  %{ endif }
  stagger           = "30s"
}
...
/```

```